### PR TITLE
Add boto3_elasticache exec and state modules

### PR DESCRIPTION
**Pull request by tkwilliams**
Adds new boto3-based modules for AWS Elasticache support.
Note that these exist in parallel to the original boto2-based versions, since they are not intended to be directly backwards compatible - the new modules use a different approach to passing through the AWS API params.
These new modules also support in-place modification of managed resources (the boto2 versions only provide for initial creation and subsequent deletion); and, most importantly they allow setting a large number of params (e.g. "SnapshotRetentionLimit", "SnapshotWindow", etc.) not exposed by boto2, and thus unsupported by the old modules.
The new modules essentially pass ALL unhandled params directly through to the underlying boto3 / AWS API layers, meaning that any new options provided at those layers will be immediately available in the salt functions (even if not yet added to the docstrings) and they thus should be largely forward-compatible.  Obviously, any entirely new functions added to boto3 would still need to be encapsulated...  As a bonus, things have been generic'ed to a high degree, so that adding new create/modify/delete functions should be trivial in most cases.

Lastly, a small amount of cleanup and re-org in the boto_secgroup exec module.

If someone will lemme know what the current 'versionadded' string should be I'll update it to suit :)

### CODE REVIEW

**Comment by tkwilliams** [comment]
PS - this is by no means complete :)

I'm pushing it to 1) make sure it merges and lints nicely, and 2) in case anyone wants to eyeball it and provide opinions...

**Comment by tkwilliams** [comment]
@cachedout  Hey Mike, not sure why the last run is showing pylint as having failed - there are no errors or warning (or aught else FTM) when I look at the pylint link?

Not that this is in any way ready to be merged - there's a fair bit to to be done on it yet before I'm satisfied - but I want to make sure I'm not missing some obscure lint errors in the process :)

Also, what would be best in the '.. versionadded' string for these new modules I'm adding?

**Comment by cachedout** [comment]
Every now and then that pylint test bails out and I'm not sure why. Re-running the tests should fix it. Sorry about that.

**Comment by cachedout** [comment]
And the version will be `Nitrogen` :]

**Comment by cachedout** [comment]
@tkwilliams Will you let me know when you think this one is ready to go? Thanks. :]

**Comment by tkwilliams** [comment]
@cachedout  There are a LOT more features I'd like to add to the modules, but what's there has working well in our environment for several days now, and I'm already getting noise from my boss to push it since it provides what WE require for the moment; so I imagine it's better to merge what's there now and add more functions later, as we (or someone else) need them....

tl;dr - please go ahead and merge it :^)

**Comment by tkwilliams** [comment]
Note that, due to infelicities in the AWS API which I simply cannot find a way to work around, state runs on boto3_elasticache.replication_group_present() ALWAYS generate an update event, even if nothing has changed.  This is because the API (and thus boto3) does not return the CURRENT values (via describe_replication_groups()) for almost all meaningful params you might set on the resource, and thus there's no way to check the current values and thus only modify what might have changed.  This ends up being a no-op on the AWS side of course (assuming nothing HAS changed) but is reported as a modify (complete with changes['old'] and changes['new'] fully populated) by salt :(

I HATE this, but (IMHO) it's better to do this than simply avoid the issue (by not supporting modifications at all) as the old elasticache module did...
